microprofile/WFLY-11529_subsystem_metrics.asciidoc

```diff
@@ -0,0 +1,145 @@
+====
+
+==== Metric name and labels
+
+The name of a WildFly metric is based on the subsystem that provides it as well as the name of the attribute from the management model.
+The name can also be prepended with a `prefix` (specified on the `/subsystem=microprofile-metrics` resource).
+The name can also be appended with the `_total` suffix if it corresponds to a Counter metrics (as explained in https://issues.jboss.org/browse/WFCORE-4238[WFCORE-4238])
+Other information are stored using labels.
+
+For example Undertow exposes a metric attribute `request-count` (which is a counter) for every Servlet in an application deployment.
+This attribute will be exposed to Prometheus with the name `wildfly_undertow_request_count_total`.
+Other information such as the name of the Servlet are added to the labels of the metric.
+
+The https://github.com/wildfly/quickstart/tree/master/helloworld-rs[helloworld-rs quickstart] creates a JAX-RS application that can be deployed in WildFly.
+ A corresponding metric will be exposed for it with the name and labels:
+
+----
+wildfly_undertow_request_count_total{deployment="helloworld-rs.war",servlet="org.jboss.as.quickstarts.rshelloworld.JAXActivator",subdeployment="helloworld-rs.war"}
+----
+
+For deployment metrics (as above), if the metric belongs to a `deployment`, the `subdeployment` label is also added (with the `deployment` label value) in
+order to facilitate aggregation of data by Prometheus.
+
+The example below returns 3 counters for Undertow servlet request count:
+* for the HTTP and HTTPS listeners in the `undertow` subsystem
+* for one servlet from the https://github.com/wildfly/quickstart/tree/master/helloworld-rs[helloworld-rs quickstart] deployment:
+----
+$ curl -v http://127.0.0.1:9990/metrics/ | grep wildfly_undertow_request_count_total
+...
+# HELP wildfly_undertow_request_count_total The number of requests this listener has served
+# TYPE wildfly_undertow_request_count counter
+wildfly_undertow_request_count_total{server="default-server",http_listener="default",} 8.0
+wildfly_undertow_request_count_total{server="default-server",https_listener="https",} 0.0
+wildfly_undertow_request_count_total{deployment="helloworld-rs.war",servlet="org.jboss.as.quickstarts.rshelloworld.JAXActivator",subdeployment="helloworld-rs.war",} 6.0
+...
+----
+
+===== Implementation Issues
+
+====== Complex WildFly Metrics Are Not Supported
+
+Resources may return arbitrarily complex return types for metrics.
+The registration code in `microprofile-smallrye-metrics` will only register WildFly metrics for simple numerical `ModelType`.
+
+====== Management Resources Added After Server Boot Will Not Expose Their Metrics
+
+The `microprofile-smallrye-metrics` extension will register any valid metric from WildFly Management Model when it is installed.
+However if other management resources are added afterwards, the extension will not be aware of them and will not register their metrics.
```

Comment:
`management resources are added afterwards` ... what can be an example of this ?

I'm expecting server reload would make them accessible by the extension.  Maybe it could be mentioned here too.
If there is less intrusive method than server reload it could be mentioned too.

Reply:
Right, if the server reloads, the new management resources would be taken into account.

However if you add a new management resource that does not require to reload the server (e.g. a new JMS Queue), the metrics for it would not be available until the next server reload.